src/prefect/backend/flow_run.py

```diff
@@ -0,0 +1,596 @@
+    with prefect.context(flow_run_id=flow_run_id):
+        with fail_flow_run_on_exception(
+            flow_run_id=flow_run_id,
+            message="Failed to execute flow: {exc}",
+        ):
+            if flow_run.flow.run_config is not None:
+                flow_state = runner_cls(flow=flow).run(**run_kwargs)
+
+            # Support for deprecated `flow.environment` use
+            else:
+                environment = flow.environment
+                environment.setup(flow)
+                environment.execute(flow)
+
+    logger.info(f"Run finished with final state {flow_state}")
+    return flow_run.update()
+
+
+@contextmanager
+def fail_flow_run_on_exception(
+    flow_run_id: str,
+    message: str = None,
+):
+    """
+    A utility context manager to set the state of the given flow run to 'Failed' if
+    an exception occurs. A custom message can be provided for more details and will
+    be attached to the state and added to the run logs. KeyboardInterrupts will set
+    the flow run state to 'Cancelled' instead and will not use the message. All errors
+    will be re-raised.
+
+    Args:
+        flow_run_id: The flow run id to update the state of
+        message: The message to include in the state and logs. `{exc}` will be formatted
+            with the exception details.
+    """
+    message = message or "Flow run failed with {exc}"
+    client = Client()
+
+    try:
+        yield
+    except KeyboardInterrupt:
+        if not FlowRun.from_flow_run_id(flow_run_id).state.is_finished():
+            client.set_flow_run_state(
+                flow_run_id=flow_run_id,
+                state=prefect.engine.state.Cancelled("Keyboard interrupt."),
+            )
+            logger.warning("Keyboard interrupt. Flow run cancelled, exiting...")
+        else:
+            logger.warning(
+                "Keyboard interrupt. Flow run is already finished, exiting..."
+            )
+        raise
+    except Exception as exc:
+        if not FlowRun.from_flow_run_id(flow_run_id).state.is_finished():
+            message = message.format(exc=exc.__repr__())
+            client.set_flow_run_state(
+                flow_run_id=flow_run_id, state=prefect.engine.state.Failed(message)
+            )
+            client.write_run_logs(
+                [
+                    dict(
+                        flow_run_id=flow_run_id,  # type: ignore
+                        name="prefect.backend.api.flow_run.execute_flow_run",
+                        message=message,
+                        level="ERROR",
+                    )
+                ]
+            )
+        logger.error(message, exc_info=True)
+        raise
+
+
+class FlowRun:
+    """
+    Object for getting information about a Flow Run from the backend API
+
+    Attributes:
+        flow_run_id: The uuid of the flow run
+        name: The name of the flow run
+        flow_id: The uuid of the flow this run is associated with
+        state: The state of the flow run
```

Comment:
what's the use case for `state` here? would `state` be better as a property?

if I'm accessing `FlowRun.state` I would expect the latest state to be queried for and returned

Reply:
`FlowRun` is intended to be an immutable view. We want to limit hitting the API as much as possible. If you want the current state you'd grab the latest data with `flow_run.update()` which creates a new instance.

Reply:
Got it, "immutable view" helped things click for me, thanks!

I'm not sure there's a good pattern to replace it, but I think the mutable nature of `task_runs` may cause issues, especially if this is intended for external use.

I'm thinking of doing something (deliberately stupid for illustrative purposes) like
```python
flow_run = create_flow_run(flow)
execute_flow_run(flow_run_id=flow_run.flow_run_id)

# okay now I want to check on my flow
flow_run.get(task=my_failed_task) # this task failed
# hmmm now I can see my task failed, did my flow run fail?
print (flow_run.state) # my flow run is Scheduled, wtf?
```

Reply:
Well that's why `task_runs` isn't actually mutable and it's just a cache. So yes you got the newest data for that task run but you don't really know that it's being stored in `flow_run`. I think the big difference here is you're not calling `flow.state()` or `flow.get_state()` you're calling `flow.state` which is presented as a _property_ which hints that it is static.

Reply:
The names should be clearer now and the behavior has been changed so only finished tasks are cached